src/physics/nonlinear_solid.cpp

```diff
@@ -132,49 +148,147 @@ void NonlinearSolid::completeSetup()
 
-    M_form->AddDomainIntegrator(new mfem::VectorMassIntegrator(rho0));
-    M_form->Assemble(0);
-    M_form->Finalize(0);
+    M_mat.reset(M->ParallelAssemble());
 
-    S_form = displacement_.createOnSpace<mfem::ParBilinearForm>();
-    S_form->AddDomainIntegrator(new mfem::VectorDiffusionIntegrator(*viscosity_));
-    S_form->Assemble(0);
-    S_form->Finalize(0);
+    C = displacement_.createOnSpace<mfem::ParBilinearForm>();
+    C->AddDomainIntegrator(new mfem::VectorDiffusionIntegrator(*viscosity_));
+    C->Assemble(0);
+    C->Finalize(0);
   }
 
-  // Set the MFEM abstract operators for use with the internal MFEM solvers
+  // We are assuming that the ODE is prescribing the
+  // acceleration value for the constrained dofs, so
+  // the residuals for those dofs can be taken to be zero.
+  //
+  // Setting iterative_mode to true ensures that these
+  // prescribed acceleration values are not modified by
+  // the nonlinear solve.
+  nonlin_solver_.nonlinearSolver().iterative_mode = true;
+  nonlin_solver_.SetOperator(residual);
+
   if (timestepper_ == serac::TimestepMethod::QuasiStatic) {
-    nonlin_solver_.nonlinearSolver().iterative_mode = true;
-    nonlinear_oper_ = std::make_unique<NonlinearSolidQuasiStaticOperator>(std::move(H_form), bcs_);
-    nonlin_solver_.SetOperator(*nonlinear_oper_);
+    // the quasistatic case is entirely described by the residual,
+    // there is no ordinary differential equation
+    residual.function = [=](const mfem::Vector& u, mfem::Vector& r) mutable {
+      H->Mult(u, r);  // r := H(u)
+      r.SetSubVector(bcs_.allEssentialDofs(), 0.0);
+    };
+
+    residual.jacobian = [=](const mfem::Vector& u) mutable -> mfem::Operator& {
+      auto& J = dynamic_cast<mfem::HypreParMatrix&>(H->GetGradient(u));
+      bcs_.eliminateAllEssentialDofsFromMatrix(J);
+      return J;
+    };
+
   } else {
-    nonlin_solver_.nonlinearSolver().iterative_mode = false;
-    timedep_oper_                                   = std::make_unique<NonlinearSolidDynamicOperator>(
-        std::move(H_form), std::move(S_form), std::move(M_form), bcs_, nonlin_solver_, *timedep_oper_params_);
-    ode_solver_->Init(*timedep_oper_);
+    // the dynamic case is described by a residual function and a second order
+    // ordinary differential equation. Here, we define the residual function in
+    // terms of an acceleration.
+    residual.function = [=](const mfem::Vector& d2u_dt2, mfem::Vector& res) mutable {
+      // TODO: we should avoid re-assembling this when not required
+      C_mat.reset(C->ParallelAssemble());
+
+      res = (*M_mat) * d2u_dt2 + (*C_mat) * (du_dt + c1 * d2u_dt2) + (*H) * (x + u + c0 * d2u_dt2);
+      res.SetSubVector(bcs_.allEssentialDofs(), 0.0);
+    };
+
+    residual.jacobian = [=](const mfem::Vector& d2u_dt2) mutable -> mfem::Operator& {
+      // J = M + c1 * C + c0 * H(u_predicted)
+      auto localJ = std::unique_ptr<mfem::SparseMatrix>(Add(1.0, M->SpMat(), c1, C->SpMat()));
```

Comment:
Separate PR of course, but it looks like there may be some utility for matrix expression templates.